isisd/isis_te.c

```diff
@@ -299,9 +354,908 @@ int isis_mpls_te_update(struct interface *ifp)
+		 sysid_print(lsp->hdr.lsp_id));
+
+	/* First parse LSP to obtain the corresponding Vertex */
+	vertex = lsp_to_vertex(ted, lsp);
+	if (!vertex) {
+		zlog_warn("Unable to build Vertex from LSP %s. Abort!",
+			  sysid_print(lsp->hdr.lsp_id));
+		return;
+	}
+
+	/* Check if Vertex has been modified */
+	if (vertex->status != SYNC) {
+		/* Vertex is out of sync: export it if requested */
+		if (IS_EXPORT_TE(mta))
+			isis_te_export(LS_MSG_TYPE_NODE, vertex);
+		vertex->status = SYNC;
+	}
+
+	/* Mark outgoing Edges and Subnets as ORPHAN to detect deletion */
+	for (ALL_LIST_ELEMENTS_RO(vertex->outgoing_edges, node, edge))
+		edge->status = ORPHAN;
+
+	for (ALL_LIST_ELEMENTS_RO(vertex->prefixes, node, subnet))
+		subnet->status = ORPHAN;
+
+	/* Process all Extended Reachability in LSP (all fragments) */
+	args.ted = ted;
+	args.vertex = vertex;
+	args.export = mta->export;
+	isis_lsp_iterate_is_reach(lsp, ISIS_MT_IPV4_UNICAST, lsp_to_edge_cb,
+				  &args);
+
+	isis_lsp_iterate_is_reach(lsp, ISIS_MT_IPV6_UNICAST, lsp_to_edge_cb,
+				  &args);
+
+	/* Process all Extended IP (v4 & v6) in LSP (all fragments) */
+	isis_lsp_iterate_ip_reach(lsp, AF_INET, ISIS_MT_IPV4_UNICAST,
+				  lsp_to_subnet_cb, &args);
+	isis_lsp_iterate_ip_reach(lsp, AF_INET6, ISIS_MT_IPV6_UNICAST,
+				  lsp_to_subnet_cb, &args);
+	isis_lsp_iterate_ip_reach(lsp, AF_INET6, ISIS_MT_IPV4_UNICAST,
+				  lsp_to_subnet_cb, &args);
+
+	/* Clean remaining Orphan Edges or Subnets */
+	if (IS_EXPORT_TE(mta))
+		ls_vertex_clean(ted, vertex, zclient);
+	else
+		ls_vertex_clean(ted, vertex, NULL);
+}
+
+/**
+ * Delete Link State Database Vertex, Edge & Prefix that correspond to this
+ * ISIS Link State PDU
+ *
+ * @param ted	Link State Database
+ * @param lsp	ISIS Link State PDU
+ */
+static void isis_te_delete_lsp(struct mpls_te_area *mta, struct isis_lsp *lsp)
+{
+	struct ls_ted *ted;
+	struct ls_vertex *vertex = NULL;
+	struct ls_node lnode = {};
+	struct ls_edge *edge;
+	struct ls_subnet *subnet;
+	struct listnode *nnode, *node;
+
+	/* Sanity Check */
+	if (!IS_MPLS_TE(mta) || !mta->ted || !lsp)
+		return;
+
+	te_debug("ISIS-TE(%s): Delete Link State TED objects from LSP %s",
+		 lsp->area->area_tag, sysid_print(lsp->hdr.lsp_id));
+
+	/* Compute Link State Node ID from IS-IS sysID ... */
+	if (lsp->level == ISIS_LEVEL1)
+		lnode.adv.origin = ISIS_L1;
+	else
+		lnode.adv.origin = ISIS_L2;
+	memcpy(&lnode.adv.id.iso.sys_id, &lsp->hdr.lsp_id, ISIS_SYS_ID_LEN);
+	lnode.adv.id.iso.level = lsp->level;
+	ted = mta->ted;
+	/* ... and search the corresponding vertex */
+	vertex = ls_find_vertex_by_id(ted, lnode.adv);
+	if (!vertex)
+		return;
+
+	te_debug("  |- Delete Vertex %s", vertex->node->name);
+
+	/*
+	 * We can't use the ls_vertex_del_all() function if export TE is set,
+	 * as we must first advertise the client daemons of each removal.
+	 */
+	/* Remove outgoing Edges */
+	for (ALL_LIST_ELEMENTS(vertex->outgoing_edges, node, nnode, edge)) {
```

Comment:
Well, yes and no. In fact, when you remove an edge, vertex or prefix, you should advertise the listener e.g. BGP-LS of the modification. And, the export function is different between ospf and isis. Thus, what we could do is to propose such common function, but providing a dedicated export function with dedicated parameters. I'm not sure that this will help to maintain the code as only OSPF and IS-IS are concern.